ticdc/ticdc-craft.md

```diff
@@ -0,0 +1,395 @@
+* [Resolved Event](#resolved-event): Represents a special time point before which the Event received is complete.
+
+## Restrictions
+
+* In most cases, the Row Changed Event of a version is sent only once, but in special situations such as node failure and network partition, the Row Changed Event of the same version might be sent multiple times.
+* On the same table, the Row Changed Events of each version which is first sent are incremented in the order of timestamps (TS) in the Event stream.
+* Resolved Events are periodically broadcasted to each MQ Partition. The Resolved Event means that any Event with a TS earlier than Resolved Event TS has been sent to the downstream.
+* DDL Events are broadcasted to each MQ Partition.
+* Multiple Row Changed Events of a row are sent to the same MQ Partition.
+
+## Benchmark
+
+* Binary protocol encodes the same events with far less amount of bytes.
+* Protobuf is a popular binary format that can be used instead of craft format. Since craft format can apply sophisticated optimizations, it can be more compact and faster than protobuf.
+
+Serialized size:
+
+| case | craft size | json size | protobuf 1 size | protobuf 2 size | craft compressed | json compressed | protobuf 1 compressed | protobuf 2 compressed |
+| :---- | :--------- | :-------- | :-------------- | :-------------- | :--------------- | :-------------- | :-------------------- | :-------------------- |
+| case 0 | 300 | 708 (136%)+ | 382 (27%)+ | 375 (25%)+ | 168 | 223 (32%)+ | 198 (17%)+ | 181 (7%)+ |
+| case 1 | 993 | 2816 (183%)+ | 1528 (53%)+ | 1482 (49%)+ | 209 | 286 (36%)+ | 235 (12%)+ | 221 (5%)+ |
+
+Encoding speed:
+
+| craft | json | protobuf 1 | protobuf 2 |
+| :---- | :--- | :--------- | :--------- |
+| 4809 ns/op | 28388 ns/op (490%)+ | 3921 ns/op (19%)- | 3645 ns/op (25%)- |
+
+Decoding speed:
+
+| craft | json | protobuf 1 | protobuf 2 |
+| :---- | :--- | :--------- | :--------- |
+| 7944 ns/op | 75822 ns/op (854%)+ | 8020 ns/op (1%)+ | 8462 ns/op (6%) |
+
+## Message format
+
+A Message contains one or more Events, arranged in the following format:
+
+Message:
+
+| uvarint | header | body | term dictionary | size tables |
+| :------ | :------ | :------ | :------ |
+| version | events header | events body | term dictionary | size tables |
+
+Header:
+
+| delta uvarint chunk | uvarint chunk | delta varint chunk | delta varint chunk | delta varint chunk |
+| :------ | :------ | :------ | :------ | :------ | :------ |
+| commit ts | event type | partition id (-1 for no partition) | schema | table |
+
+Body for [Row Changed Event](#row-changed-event):
+
+| Column group | Column group |
+| :------ | :------ |
+| column group 1 | column group 2 (optional) |
+
+Column group:
+| 1 byte | uvarint | delta varint chunk | uvarint chunk | uvarint chunk | nullable bytes chunk |
+| :------ | :------ | :------ | :------ | :------ | :------ |
+| type: 1 New, 2 Old | number of columns | name | type | flag | value |
```

Comment:
Hm, that makes sense. But unfortunately the code was merged and a version was released. We can't break binary compatibility without a codec version upgrade. 

Try to recall why I made it like this. Since old value might be missing, and new value is mandatory therefore I make it like so.